README.md
# Rails Excel

It adds support of .rxls templates for your rails views

[DEPRECATED]

Prior to 1.0.0, it comes with two builtin strategies based on the very good gems [writeexcel](https://github.com/cxn03651/writeexcel) and [spreadsheet](http://spreadsheet.rubyforge.org/index.html)

[IMPORTANT]

As of 1.0.0, there is no builtin strategy anymore. You should install at least one strategy :

* [spreadsheet](https://github.com/hallelujah/rails-excel-spreadsheet-strategy)
* [write_excel](https://github.com/hallelujah/rails-excel-writeexcel-strategy)
* [rubyXL](https://github.com/hallelujah/rails-excel-rubyXL-strategy)



# Requirements

* rails >= 2.3.0
* ruby >= '1.8.6'


# Usage

In your config/environment.rb

```ruby
config.gem 'rails-excel'
```

Create an initializer : config/initializers/excel.rb

```ruby
RailsExcel.configure do |config|
  config.strategy = :spreadsheet # by default or :write_excel
end
```

If you wan tot implement your own strategy, here are the requirements :

* Must respond to compile
* First argument of compile takes a StringIo instance where to write the response
* compile takes a block that yields a workbook instance


Example extracted from code source:

```ruby
# lib/my_strategy.rb

class MyStrategy
  def compile(io, &block)
    workbook = ::Spreadsheet::Workbook.new
    yield(workbook)
    workbook.write(io)
  end
end
```

Then in your config/initializers/excel.rb

```ruby
require 'my_strategy'
RailsExcel.configure do |config|
  config.add_strategy :my_strategy, MyStrategy.new
  # Redefining default strategy
  config.strategy = :my_strategy # by default it was :spreadsheet
end
```

You can use any object as long as it responds to described `compile` method

The strategy defined in the initializer will be used in all of your controllers

To use another strategy you can set it by controller :

```ruby
class UsersController < ApplicationController
  self.excel_strategy = :write_excel
end
```

Or you can set strategy per action by redefining instance method excel_strategy

```ruby
class UsersController < ApplicationController
  self.excel_strategy = :write_excel
  def index
    # ...
  end

  def show
    # ...
  end

  protected
  def excel_strategy
    case action_name
    when 'index'
      :spreadsheet
    else
  self.class.excel_strategy
    end
  end
end
```

Or you can also override excel_strategy by action like that :

```ruby
class UsersController < ApplicationController
  self.excel_strategy = :write_excel

  def other_action
    self.excel_strategy = :rubyXL
    # ...
  end

end
```

# Contributors

* [edpaget](https://github.com/edpaget)


Changelog
2012-01-02  Hallelujah - Ramihajamalala Hery <hery@rails-royce.org>

	* Fix @available_strategies not set : v1.0.2
	* Fix requiring of strategies : v1.0.1
	* Remove builtin strategies : v1.0.0
	* Deprecate Rails::Excel in favor of RailsExcel v0.0.4

2011-09-26 18:33  Ramihajamalala Hery <hery@rails-royce.org> v0.0.3

	* Compatibility with Rails 3.x (by edpaget)

2011-08-23 12:19  Ramihajamalala Hery <hery@rails-royce.org> v0.0.2

	* Add a todo list
	* Allow to set excel strategy per action basis

2011-08-22 10:19  Ramihajamalala Hery <hery@rails-royce.org> v0.0.1

	* First working release



lib/rails-excel.rb
require 'activesupport'
require "rails-excel/version"
require 'rails-excel/delegation'
require 'rails-excel/template_handler'

module RailsExcel
  @available_strategies = Hash.new

  class << self
    module_eval do

      attr_accessor :strategy
      attr_reader :available_strategies

      def add_strategy(name, instance)
        strategy = instance
        if Class === instance && !instance.respond_to?(:compile)
          strategy = instance.new
        end
        @available_strategies[name.to_sym] = strategy
      end

      def configure(&block)
        yield(self)
        ::ActionView::Base.send :include, RailsExcel::Delegation::View
        ::ActionController::Base.send :include,RailsExcel::Delegation::Controller
        ActionView::Template.register_template_handler :rxls, RailsExcel::TemplateHandler
      end

    end
  end

end

module Rails
  module Excel
    class << self
      module_eval do
        def add_strategy(name, instance)
          ActiveSupport::Deprecation.warn("Rails::Excel is deprecated in favor of RailsExcel",caller)
          RailsExcel.add_strategy(name, instance)
        end
        def configure(&block)
          ActiveSupport::Deprecation.warn("Rails::Excel is deprecated in favor of RailsExcel",caller)
          RailsExcel.configure(&block)
        end
      end
    end

  end
end




lib/rails-excel/strategies/spreadsheet.rb
require 'spreadsheet'


module RailsExcel
  module Strategies

    class Spreadsheet
      def compile(io, &block)
        workbook = ::Spreadsheet::Workbook.new
        yield(workbook)
        workbook.write(io)
      end
    end

  end

end

# This extends colors for spreadsheet gem
# See http://dmcritchie.mvps.org/excel/colors.htm#dpalette

colors = %w{
black brown olive_green dark_green dark_teal dark_blue indigo dark_gray
dark_red orange dark__yellow green teal blue blue_gray gray
red light_orange lime sea_green aqua light_blue violet light_gray
pink gold yellow righ_green turquoise sky_blue plu lite_gray
rose tan light_yellow light_green light_turquoise plae_blue lavender white
periwinkle plum_plus ivory lite_turquoise dark_purple coral ocean_blue ice_blue
dark_blue pink_plus yellow_plus turquoise violet dark_red teal_plus blue_plus
}

color_indexes = %w{
1 53 52 51 49 11 55 56
9 46 12 10 14 5 47 16
3 45 43 50 42 41 13 48
7 44 6 4 8 33 54 15
38 40 36 35 34 37 39 2
17 18 19 20 21 22 23 24
25 26 27 28 29 30 31 32
}

color_indexes.each_with_index do |c,i|
  # Do not override already defined colors
  color_num = c.to_i + 7
  Spreadsheet::Excel::Internals::COLOR_CODES[color_num] = colors[i].to_sym unless Spreadsheet::Excel::Internals::COLOR_CODES.key?(color_num)
  Spreadsheet::Excel::Internals::SEDOC_ROLOC[colors[i].to_sym] = color_num unless Spreadsheet::Excel::Internals::COLOR_CODES.key?(colors[i].to_sym)
end

# Add palette colors but not override existing colors
ObjectSpace.each_object(Class)  do |m|
  if m.ancestors.include?(Spreadsheet::Datatypes)
    m::module_eval do
      class << self
        COLORS.replace(COLORS | Spreadsheet::Excel::Internals::SEDOC_ROLOC.keys)
      end
    end
  end
end
